README.md
<!-- badges: start -->
[![CRAN
Version](https://img.shields.io/cran/v/lodi?style=flat-square&color=blue&label=CRAN)](https://cran.r-project.org/package=lodi)
[![GitHub
Release](https://img.shields.io/github/v/release/umich-cphds/lodi?include_prereleases&label=Github&style=flat-square)](https://github.com/umich-cphds/lodi)
[![Travis
CI](https://img.shields.io/travis/umich-cphds/lodi?style=flat-square)](https://travis-ci.org/umich-cphds/lodi)
<!-- badges: end -->

Limit of Detection Multiple Imputation
======================================

`lodi` is a R package that implements censored likelihood multiple
imputation (CLMI) for single pollutant models with exposure biomarkers
below their respective detection limits. `lodi` also contains
implementations for standard methods such as single imputation with a
constant and complete-case analysis, although those methods are
primarily designed for comparison with `clmi`.

Installation
------------

`lodi` requires `rlang >= 0.3.0` to be installed, so you may want to
install or update `rlang` before installing `lodi`.

The package can be installed from CRAN

    install.packages("lodi")

Or from Github

    # install.packages("devtools")
    devtools::install_github("umich-cphds/lodi", build_opts = c())

The Github version may contain bug fixes not yet present on CRAN, so if
you are experiencing issues, you may want to try the Github version of
the package.

Example
-------

Once `lodi` is installed, you can load up R and type

    vignete("lodi")

to learn how to use the method.

Bugs
----

If you encounter a bug, please open an issue on the
[Issues](https://github.com/umich-cphds/lodi/issues) tab on Github or
send us an email.

Contact
-------

For questions or feedback, please email Jonathan Boss at
<bossjona@umich.edu> or Alexander Rix <alexrix@umich.edu>.

References
----------

Boss J, Mukherjee B, Ferguson KK, et al. Estimating outcome-exposure
associations when exposure biomarker detection limits vary across
batches. *Epidemiology*. 2019;30(5):746-755.
[10.1097/EDE.0000000000001052](https://doi.org/10.1097/EDE.0000000000001052)


NEWS.md
# lodi v1.0.0

## Major changes
* lodi now requires R >= 3.5.0 due to object serialization changes
 
# lodi v0.9.2

## Minor changes
* Added badges to README.Rmd / README.md to show the CRAN version of lodi, Github version of lodi, and a Travis-CI badge showing whether or not lodi tests are passing. If the Github version and CRAN versions differ, the badges appear as different colors.
* Added some unit tests to be run when updating the package / before CRAN submission.
* Added check to ensure exposure data that is below the `lod` is coded as `NA`. Users might input an already log transformed exposure in clmi, while not transforming the `lod` variable. If the transformation is logarithmic, this will catch it
* Changed the optimization algorithm uses in the MLE procedure from `L-BFGS-B` to `BFGS` with reparameterized variance (ie  `exp(var)`) to ensure non-negativity instead of box constraints. Using `L-BFGS-B` could cause non-finite values of the objective function due to large gradients when evaluating the initial value of the function.

## Bug fixes
* Fixed a subtle bug when calling lodi inside of a function when using a custom exposure function. lodi was internally using the parent frame to enclose the transformation, which is the lodi package namespace, instead of the caller's frame.

# lodi v0.9.1

## Minor changes
* Added additional checks to `clmi` to protect against improper variable ordering.
* Added `verbose` option to `clmi` to print out parsing information.
* Updated `clmi` documentation to provide additional exposition on proper use of the package.

## Bug fixes
* Corrected calculation of censored likelihood maximum likelihood estimates, and named them.
* `clmi` now returns imputed data.frames that have the same row ordering as the original.

# lodi v0.9.0
Initial CRAN release.


tests/testthat.R
library(testthat)
library(lodi)

test_check("lodi")


tests/testthat/test_clmi.R
context("Test clmi")


test_that("clmi throws errors correctly", {
    lod.var <- toy_data$lod
    expect_error(clmi("poll ~ case_cntrl + smoking + gender, toy_data", lod, 1))
    expect_error(clmi(poll ~ case_cntrl + smoking + gender, toy_data, lodu, 1))
    expect_error(clmi(poll ~ case_cntrl + smoking + gender, toy_data, NULL, 1))

    expect_error(clmi(a * poll ~ case_cntrl + smoking + gender, toy_data,
                      lod, 1))

    a <- 2
    fn <- function(x) a * x
    expect_true(!is.null(clmi(fn(poll) ~ case_cntrl + smoking + gender, toy_data,
                      lod, 1)))

    expect_error(clmi(poll ~ case_cntrl + smoking + gender, "toy_data", lod, 1))
    expect_error(clmi(poll ~ case_cntrl + smoking + gender, toy_data, lod, "a"))
    expect_error(clmi(poll ~ case_cntrl + smoking + gender, toy_data, lod, 1, 0))
    expect_error(clmi(poll ~ case_cntrl + smoking + gender, toy_data, lod,
                      1, c(2, 4)))

    df <- toy_data
    df$lod <- NULL
    expect_error(clmi(poll ~ case_cntrl + smoking + gender, df, lod, 1))

    # Test for improper lod values
    df <- toy_data
    df$poll <- log(df$poll)
    expect_error(clmi(poll ~ case_cntrl + smoking + gender, df, lod, 1))

    # Test that missing covariates throw an error
    df <- toy_data
    df$smoking[1] <- NA
    expect_error(clmi(poll ~ case_cntrl + smoking + gender, df, lod, 1))

    # Test that factors cause an error
    df <- toy_data
    df$gender <- as.factor(df$gender)
    expect_error(clmi(poll ~ case_cntrl + smoking + gender, df, lod, 1))
})


R/cca-r2.R
#' Single pollutant complete case analysis.
#'
#' lod_cca is a helper function that does complete case analysis for
#' single pollutant models.  The function can be used to compare with clmi.
#' @param formula A R formula in the form outcome ~ exposure + covariates.
#' @param df A data.frame that contains the variables \code{formula}
#'  references.
#' @param type The type of regression to perform. Acceptable options are
#'   linear and logistic.
#' @examples
#' library(lodi)
#' # load lodi's toy data
#' data("toy_data")
#' x <- lod_cca(case_cntrl ~ poll + smoking + gender, toy_data, logistic)
#' # see the fit model
#' x$model
#' @export
lod_cca <- function(formula, df, type)
{
  if (class(formula) != "formula")
      stop("formula must be a formula")
  if (!is.data.frame(df))
        stop("df must be a data.frame")

  # get the type of regression to run
  type <- deparse(substitute(type))
  type <- match.arg(type, c("linear", "logistic"))

  df <- stats::na.omit(df)

  mod.fr <- stats::model.frame(formula, data = df)
  if (type == "linear")
    model <- stats::lm(mod.fr)
  else
    model <- stats::glm(mod.fr, family = stats::binomial())
  list(model = model, formula = formula, data = df)
}

#' Single pollutant \code{sqrt(2)} imputation.
#'
#' lod_root2 is a helper function that performs single imputation with
#' \code{lod / sqrt(2)}, a common ad hoc approach used in single-pollutant
#' modeling. The function can be used to compare with clmi.
#' @param formula A R formula in the form \code{outcome ~ exposure + covariates}.
#' @param df A data.frame that contains the variables \code{formula}
#'    references.
#' @param lod Name of the limit of detection variable.
#' @param type The type of regression to perform. Acceptable options are
#'     linear and logistic.
#' @note
#'     Depending on the transformation used, a "Complicated transformation"
#'     error may occur. For example, the transformation \code{a * exposure} will
#'     cause an error. In this case, define a transformation function as
#'     \code{f <- function(exposure) a * exposure} and use \code{f} in your
#'     formula. This technical limitation is unavoidable at the moment.
#' @examples
#' # load lodi's toy data
#' library(lodi)
#' data("toy_data")
#' lodi.out <- lod_root2(case_cntrl ~ poll + smoking + gender, toy_data, lod,
#'                         logistic)
#' # see the fit model
#' lodi.out$model
#'
#' # we can log transform poll to make it normally distributed
#' lodi.out <- lod_root2(case_cntrl ~ log(poll) + smoking + gender, toy_data,
#'                         lod, logistic)
#' lodi.out$model
#'
#' # transforming the exposure results in a new column being added to data,
#' # representing the transformed lod.
#' head(lodi.out$data)
#'
#' # You can even define your own transformation functions and use them
#' f <- function(x) exp(sqrt(x))
#' lodi.out <- lod_root2(case_cntrl ~ f(poll) + smoking + gender, toy_data, lod,
#'                         logistic)
#' head(lodi.out$data)
#' @export
lod_root2 <- function(formula, df, lod, type)
{
    if (class(formula) != "formula")
        stop("formula must be a formula")
    if (!is.data.frame(df))
        stop("df must be a data.frame")

    # get the first term on the rhs formula, an expression involving 'exposure'.
    transform.init <- rlang::parse_expr(labels(stats::terms(formula))[1])

    # get the exposure from the transformation
    exposure  <- all.vars(transform.init)
    if (length(exposure) > 1)
        stop("Complicated transformation on exposure. See help for fix.")


    # create a variable and assign it the symbol 'x'.
    assign(exposure, quote(x))

    # get the transformation and replace exposure with the symbol x.
    transform <- eval(substitute(substitute(transform.init)))

    # Calculate the transformation function
    t.function <- function(x) x
    body(t.function) <- transform
    environment(t.function) <- rlang::caller_env()

    # calculate the name of the transformed lod variable
    assign("x", substitute(lod))
    transform.lod <- deparse(eval(substitute(substitute(transform))))

    lod <- deparse(substitute(lod))
    if (is.null(df[[lod]]))
        stop(paste(lod, "not in data."))

    # get the type of regression to run
    type <- deparse(substitute(type))
    type <- match.arg(type, c("linear", "logistic"))

    # impute NAs as lod / sqrt(2)
    tmp <- df[[exposure]]
    for (i in 1:nrow(df))
        tmp[i] <- ifelse(is.na(tmp[i]), df[[lod]][i] / sqrt(2), tmp[i])

    df[[exposure]] <- tmp
    df[[transform.lod]] <- t.function(df[[lod]])
    df[[deparse(transform.init)]] <- t.function(df[[exposure]])

    mod.fr <- stats::model.frame(formula, data = df)
    if (type == "lm")
        model <- stats::lm(mod.fr)
    else
        model <- stats::glm(mod.fr, family = stats::binomial())

    list(model = model, formula = formula, data = df, t.function = t.function)
}


R/clmi.R
#' Censored Likelihood Multiple Imputation
#'
#' This function performs censored likelihood multiple imputation for
#' single-pollutant models where the pollutant of interest is subject to
#' varying detection limits across batches (this function will also work if
#' there is only one distinct detection limit). The function
#' outputs a list containing the imputed datasets and details regarding the
#' imputation procedure (i.e., number of imputed dataset, covariates used to
#' impute the non-detects, etc).
#'
#' \code{clmi} is somewhat picky regarding the \code{formula} parameter. It
#' tries to infer what transformation you'd like to apply to the exposure you
#' are imputing, what the exposure is, and what the outcome is. It attempts to
#' check to make sure that everything is working correctly, but it can fail.
#' Roughly, the rules are:
#' \itemize{
#'   \item The left hand side of formula should be the exposure you are trying
#'     to impute.
#'   \item The exposure may be optionally wrapped in a univariate transformation
#'     function. If the transformation function is not univariate, you ought to
#'     get an error about a "complicated" transformation.
#'   \item The first variable on the right hand side of \code{formula} should be
#'     your outcome of interest.
#'}
#' @param formula A formula in the form of \code{exposure ~ outcome + covariates}.
#' That is, the first variable on the right hand side of \code{formula} should
#' be the outcome of interest.
#' @param df A data.frame with \code{exposure}, \code{outcome} and
#'   \code{covariates}.
#' @param lod Name of limit of detection variable in \code{df}.
#' @param n.imps Number of datasets to impute. Default is 5.
#' @param seed For reproducability.
#' @param verbose If \code{TRUE}, \code{clmi} prints out useful debugging
#'   information while running. Default is \code{FALSE}.
#' @note
#' \itemize{
#'   \item \code{clmi} only supports categorical variables that are numeric,
#'     (i.e., not factors or characters). You can use the \code{model.matrix}
#'     function to convert a data frame with factors to a numeric design matrix
#'     and subsequently convert that matrix back into a data frame using
#'     \code{as.data.frame}.
#'   \item If you get the error message "L-BFGS-B needs finite values of 'fn'",
#'     try normalising your data.
#' }
#' @examples
#' library(lodi)
#'
#' # Note that the outcome of interest is the first variable on the right hand
#' # side of the formula.
#' clmi.out <- clmi(poll ~ case_cntrl + smoking + gender, toy_data, lod, 1)
#'
#' # you can specify a transformation to the exposure in the formula
#' clmi.out <- clmi(log(poll) ~ case_cntrl + smoking + gender, toy_data, lod, 1)
#'
#' @references
#'   Boss J, Mukherjee B, Ferguson KK, et al. Estimating outcome-exposure
#'   associations when exposure biomarker detection limits vary across batches.
#'   Epidemiology. 2019;30(5):746-755.
#'   \href{https://doi.org/10.1097/EDE.0000000000001052}{10.1097/EDE.0000000000001052}
#' @export
clmi <- function(formula, df, lod, seed, n.imps = 5, verbose = FALSE)
{
    if (!rlang::is_formula(formula))
        stop("formula must be a formula")
    if (!is.data.frame(df))
        stop("df must be a data.frame.")
    if (!is.logical(verbose))
        stop("'verbose' must take on a logical value.")

    if (verbose)
        print(paste("Formula:", rlang::expr_text(formula)))

    transform.init <- rlang::f_lhs(formula)
    exposure <- all.vars(transform.init)
    if (length(exposure) > 1)
        stop("Complicated transformation on exposure. See help for fix.")

    if (verbose)
        print(sprintf("Exposure variable: %s", exposure))

    vars <- all.vars(rlang::f_rhs(formula))
    outcome <- vars[1]
    if (verbose)
        print(sprintf("Outcome variable: %s", outcome))

    # Calculate the transformation function
    assign(substitute(exposure), quote(x))
    transform <- eval(rlang::expr(substitute(!!transform.init)))
    trans <- function(x) x
    rlang::fn_body(trans) <- transform
    rlang::fn_env(trans) <- rlang::caller_env()

    if (verbose)
        print(sprintf("Transformation function: %s",
              gsub("\n", "", rlang::expr_text(trans))
    ))

    lod <- deparse(substitute(lod))
    if (verbose)
        print(sprintf("LOD variable: %s", lod))

    if (is.null(df[[lod]]))
        stop(sprintf("%s not in df", lod))

    if (!is.numeric(df[[lod]]))
        stop(sprintf("%s must be numeric."))

    if (!is.numeric(seed) || length(seed) > 1)
        stop("seed must be a number.")

    set.seed(seed)
    if (!is.numeric(n.imps) || length(n.imps) > 1)
        stop("n.imps must be an integer.")
    if (n.imps < 1)
        stop("n.imps must be >= 1")

    vars <- c(exposure, vars)
    if (any(sapply(vars, function(x) !is.numeric(df[[x]]))))
        stop("clmi only supports floating point / integer variables.")
    if (any(sapply(vars[-1], function(x) any(is.na(df[[x]])))))
        stop("The covariates on the rhs of formula cannot contain missing values.")


    if (any(stats::na.omit(df[[exposure]] < df[[lod]])))
        stop(paste(exposure, "contains values below", lod,
                   "that are not coded as NA."))

    # columns to add back at the end
    leftovers <- df[, setdiff(names(df), c(vars,  lod))]

    # will be used to ensure the imputed column ordering is the same
    df.names    <- names(df)
    df.rownames <- rownames(df)

    df <- df[, c(vars, lod)]
    t.imp.exp <- paste0(exposure, "_transform", "_imputed")

    df[[t.imp.exp]] <- df[[exposure]]

    obs.above <- !is.na(df[[t.imp.exp]])
    # Subjects with concentration above LOD
    above <- df[obs.above,]
    # Subjects with concentration below LOD for each batch
    below <- df[!obs.above,]

    above.matrix <- as.matrix(above[, vars[-(1:2)]])
    below.matrix <- as.matrix(below[, vars[-(1:2)]])

    # Perform Multiple Imputation
    imp <- rep(list(below), n.imps)
    for (j in 1:n.imps) {
        #Bootstrap data
        df.bs <- df[sample(nrow(df), nrow(df), T),]

        above.bs <- df.bs[!is.na(df.bs[[t.imp.exp]]),]
        above.bs[[t.imp.exp]] <- trans(above.bs[[t.imp.exp]])

        below.bs <- df.bs[is.na(df.bs[[t.imp.exp]]),]
        below.bs[[lod]] <- trans(below.bs[[lod]])

        above.matrix.bs <- as.matrix(above.bs[, vars[-(1:2)]])
        below.matrix.bs <- as.matrix(below.bs[, vars[-(1:2)]])

        # calculates the means of (1, exposure, covariates) given beta
        mu <- function(beta, outcome, covars)
        {
            beta[2] + beta[3] * outcome + covars %*% beta[-(1:3)]
        }

        # objective function for mle is smooth and convex. So, it has a unique
        # global minimum. `sigma^2` is parameterized as beta[1] = log(sigma^2),
        # to add a non negativity constraint.
        objective <- function(beta)
        {
            mu.b <- mu(beta, below.bs[[outcome]], below.matrix.bs)
            mu.a <- mu(beta, above.bs[[outcome]], above.matrix.bs)
            -sum(log(stats::pnorm(below.bs[[lod]], mu.b, sqrt(exp(beta[1]))))) +
            sum((above.bs[[t.imp.exp]] - mu.a)^2) / (2 * exp(beta[1])) +
            nrow(above.bs) * 0.5 * log(2 * pi * exp(beta[1]))
        }

        # get MLE for Bootstrapped sample
        beta <- c(0, 0, 0, rep(0, length(vars[-(1:2)])))

        mle <- stats::optim(beta, objective, method = "BFGS")

        # Impute missing values
        mus   <- mu(mle$par, below[[outcome]], below.matrix)
        sigma <- sqrt(exp(mle$par[1]))

        normalize     <- function(x, mu, sd) (x - mu) / sd
        inv.normalize <- function(x, mu, sd) x * sd + mu

        probs <- stats::pnorm(normalize(trans(below[[lod]]), mus, sigma))
        zs    <- stats::qnorm(stats::runif(nrow(below)) * probs)
        vals  <- inv.normalize(zs, mus, sigma)
        imp[[j]][[t.imp.exp]] <- as.vector(vals)
    }

    # Check MLE estimation for original dataset
    mu <- function(beta, outcome, covars)
    {
        beta[2] + beta[3] * outcome + covars %*% beta[-(1:3)]
    }

    objective <- function(beta)
    {
        mu.b <- mu(beta, below[[outcome]], below.matrix)
        mu.a <- mu(beta, above[[outcome]], above.matrix)
        -sum(log(stats::pnorm(trans(below[[lod]]), mu.b, sqrt(exp(beta[1]))))) +
        sum((trans(above[[t.imp.exp]]) - mu.a)^2) / (2 * exp(beta[1])) +
        nrow(above) * 0.5 * log(2 * pi * exp(beta[1]))
    }

    mle <- stats::optim(beta, objective, method = "BFGS", hessian = T)

    param    <- mle$par
    param[1] <- exp(param[1])

    names(param) <- c("variance", "intercept", outcome, vars[-(1:2)])

    # due to the reparameterization of sigma^2, we need to multiply the fisher
    # information matrix by the jacobian.
    jacobian <- diag(c(param[1], 1, 1, rep(1, length(vars[-(1:2)]))))
    fisher.inf <- jacobian %*% solve(mle$hessian) %*% jacobian
    prop.sigma <- sqrt(diag(fisher.inf))

    #Aggregate imputed datasets with observed dataset
    imputed.dfs <- imp
    above[[t.imp.exp]] <- trans(above[[t.imp.exp]])

    # add imputed values, then add back the rest of the columns to to imputed
    # dataframe and make sure it is in the same order as the original
    imputed.dfs <- lapply(imputed.dfs, function(df)
    {
        df <- rbind(df, above)
        # reorder the rows so they match the original
        df <- df[df.rownames, ]
        df <- cbind(df, leftovers)
        # reorder the columns so they match the original
        df[, c(df.names, t.imp.exp)]

    })

    structure(
        list(formula = formula, nimp = n.imps, imputed.dfs = imputed.dfs,
            t.function = trans, par.mle = param, fisher.inf = fisher.inf),
        class = "clmi.out")
}

#' Calculate pooled estimates from \code{clmi.out} objects using Rubin's rules
#' @param formula Formula to fit. Exposure variable should end in
#'     \code{_transform_imputed}.
#' @param clmi.out An object generated by clmi.
#' @param type Type of regression to pool. Valid types are
#'     logistic and linear.
#' @examples
#' # continue example from clmi
#' # fit model on imputed data and pool results
#' library(lodi)
#' data("toy_data")
#' clmi.out <- clmi(log(poll) ~ case_cntrl + smoking + gender, toy_data, lod, 1)
#' results <- pool.clmi(case_cntrl ~ poll_transform_imputed + smoking, clmi.out,
#'                        logistic)
#'
#' results$output
#' @export
pool.clmi <- function(formula, clmi.out, type)
{
    if (class(clmi.out) != "clmi.out")
        stop("clmi.out is not an clmi.out object.")

    type <- deparse(substitute(type))
    type <- match.arg(type, c("linear","logistic"))

    # Get names used in clmi() function
    imputed.dfs <- clmi.out$imputed.dfs

    n.imps <- clmi.out$nimp
    nobs <- nrow(imputed.dfs[[1]])

    # Get number of coefficients to store
    # Get estimates and standard errors for each imputed dataset
    betas       <- c()
    varcov      <- vector("list", n.imps)
    regressions <- vector("list", n.imps)

    if (type == "logistic") {
        for(i in 1:n.imps) {
            data.fit <- imputed.dfs[[i]]
            logreg <- stats::glm(formula, family = "binomial", data = data.fit)
            betas <- rbind(betas, stats::coefficients(logreg))
            varcov[[i]]      <- summary(logreg)$cov.unscaled
            regressions[[i]] <- logreg
        }
    } else {
        for (i in 1:n.imps) {
            data.fit <- imputed.dfs[[i]]
            linreg <- stats::lm(formula, data = data.fit)
            betas <- rbind(betas, stats::coefficients(linreg))
            varcov[[i]]      <- stats::vcov(linreg)
            regressions[[i]] <- linreg
        }
    }

    # Pooled Inference for Multiply Imputed Datasets
    qbar <- apply(betas, 2, mean)
    ubar <- Reduce('+', varcov) / n.imps

    # Each column of this matrix is Qi - Qbar
    q_diff <- apply(betas, 1, function(x) x - qbar)
    ui <- matrix(0, ncol(betas), ncol(betas))
    for (m in 1:ncol(q_diff))
        ui <- ui + as.matrix(q_diff[,m]) %*% t(as.matrix(q_diff[,m]))

    b <- ui / (n.imps-1)
    t <- ubar + (1 + (1 / n.imps)) * b
    gamma <- ((1 + (1 / n.imps)) * diag(b)) / diag(t)
    r <- ((1 + (1 / n.imps)) * diag(b)) / diag(ubar)
    v <- (n.imps - 1) * (1 + (1 / r))^2
    v0 <- nobs - ncol(betas)
    denom_adj_df <- ((1 - gamma) * v0 * (v0 + 1)) / (v0 + 3)
    vstar <- ((1 / v) + (1 / denom_adj_df))^(-1)

    #Get Confidence Intervals
    LCL <- qbar - stats::qt(0.975, vstar) * sqrt(diag(t))
    UCL <- qbar + stats::qt(0.975, vstar) * sqrt(diag(t))

    #Get p-values
    p_vals <- 2 * stats::pt(abs(qbar / sqrt(diag(t))), df = vstar,
                                lower.tail = F)

    #Summary of each regression models on each imputed datasets
    regression.summaries <- lapply(regressions, summary)

    list(output = data.frame(est = qbar, se = sqrt(diag(t)), df = vstar,
        p.values = p_vals, LCL.95 = LCL, UCL.95 = UCL),
        pooled.vcov = t, regressions = regressions,
        regression.summaries = regression.summaries
    )
}
